qiskit/opflow/primitive_ops/pauli_sum_op.py

```diff
@@ -113,10 +113,18 @@ def add(self, other: OperatorBase) -> OperatorBase:
                 f"{other.num_qubits}, is not well defined"
             )
 
-        if isinstance(other, PauliSumOp):
+        if (
+            isinstance(other, PauliSumOp)
+            and isinstance(self.coeff, (int, float, complex))
+            and isinstance(other.coeff, (int, float, complex))
```

Comment:
Does this cover all numeric types? Wouldn't it be safer to check the other way around, if it is not a `ParameterExpression`?

Reply:
As you can see from the type hints, we do not support other types. If we need support, should I rewrite everything else to support numpy.integer, floating, complexfloating (e.g. in OperatorBase.mul())?

Reply:
Technically you're correct, but I think we can be a bit more accommodating for numeric types and support all kinds of numbers. I think when we say "we support (int, float, complex)" that implicitly stands for any number. Sometimes users accidentally use a numpy type without knowing it (or I do 😄). 

Thus it's much safer to check the other way around, for a `ParameterExpression`.

Reply:
If the `_multiply` method in `SparsePauliOp` will accept any `Number`, then we should allow sending any `Number`, right? 
Isn't this independent of the specification of `OperatorBase.mul`?

I support the liberal route, checking for `ParameterExpression`. (I did this in the fix I provided for the user with the problem: https://github.com/jlapeyre/qiskit-terra/tree/pauli-sum-op-add)

Or maybe better, checking if we are sending a `Number`, since that is what `SparsePauliOp._multiply` claims to support.

Reply:
OK, we don't officially support numpy's numeric types for now. I also prefer much safer way. So I will follow your suggestion.

@jlapeyre Thanks. Sorry, my wording was bad. `PrimitiveOps.coeff` is assumed to be `int, float, complex or ParameterExpression`. So, passing numpy's numeric types becomes undefined behavior (it may be work).